### PR TITLE
gateway-api: Add opt-in validating webhook

### PR DESCRIPTION
**Pull request by howardjohn**
This adds the ability to _opt-in_ to validating gateway-api types. This is off by default.

There are a variety of changes around making validation server robust to non-Istio types, which is why there is seemingly unrelated changes.

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
why not leave the validation to the gateway api itself? Once we opt-in, this may break compatibility because of gateway API itself is iterating fast, for this we can see the istio gateway controller is changing when the gateway API bumps. On the other aspect,  the responsibility of validation is better to be in upstream instead of each downstream meshes

**Comment by howardjohn** [comment]
> why not leave the validation to the gateway api itself? Once we opt-in, this may break compatibility because of gateway API itself is iterating fast, for this we can see the istio gateway controller is changing when the gateway API bumps. On the other aspect, the responsibility of validation is better to be in upstream instead of each downstream meshes

Good question. I thought the same initially and changed my mind, so let me explain my thoughts.

First, we *have* to have a webhook somehow; this is enforced in conformance soon. So we have to get it from somewhere.

A few options:

- the Kubernetes platform provides it. This will never be core probably, but I expect things like GKE and OpenShift may come with the CRD and Validation out of the box. This is the ideal situation, but not universal. 
- Upstream provides it. They do have an implementation. Or some other implementation provides it. 
- Istio provides it. The benefit of this is that maintaining a webhook is a lot of operational burden. You need to install+upgrade it, manage certificates, have high rbac privilege to patch the caBundle, etc - the example provided by upstream is not really production ready either. Istio already had this cost paid for the Istio webhook, so adding gateway validation on top of that adds zero incremental cost.

Basically, this gives the option for users to get a webhook server for free, when otherwise it may be very expensive. It's *off by default* because it's not the ideal solution (which is the platform providing it).

I am not sure about the compatibility issues. If users opt in to the webhook they are opting into the API version associated with that version of Istio. We can clearly document that this is the case and that we follow the upstream API so any compatibility issues are dependent on upstream, not Istio decisions. This compatibility problem exists regardless of if Istio is validating or anything else, so it doesn't really matter *where* the webhook runs; breaking changes in it will impact the user, so the API ought to not make breaking changes.

One other note is this is the recommendation of the gateway-api project to offer an option like this.

**Comment by howardjohn** [comment]
@hzxuzhonghu lmk if that makes sense

**Comment by hzxuzhonghu** [comment]
I am still feeling a bit hard to maintain it, some times users has installed gateway crd  `version x` beforehand, and then install istio 1.16, with gateway API validation say `version y`.  Let's say x and y are compatible.

Later user may upgrade to istio 1.17 with gateway API support for `version z`, there is a great risk that gateway API can not keep compatible.  Istio would better not to tightly coupled with gateway API until it guarantee compatibility like k8s API currently.

**Comment by howardjohn** [comment]
While that is true, CRD validation can only become more lenient, not more strict for backwards compat. So the risk would be that a user has a older Istio vesion than CRD and its a bit more strict, but that is fine - Istio itself probably relies on that strictness, so its fine to be more strict. When users upgrade istio they can then create the policies that may be denied.

Note this is **opt in** - users should only enable this if they only use gateway for Istio (or istio is their primary use case). In that case, they likely want validation in the context of Istio.

Ultimately, we _have_ to have validation to be conformant, and this is the officially recommended approach from upstream, so I feel like this is still the best path forward. If we were making it on-by-default I would agree we should wait until `v1` of the API probably, but for now this seems fairly safe?

File: operator/pkg/apis/istio/v1alpha1/values_types.proto [hunk 1]
```diff
@@ -1130,6 +1130,7 @@ message BaseConfig {
   // For istioctl usage to disable istio config crds in base
   google.protobuf.BoolValue enableIstioConfigCRDs = 3;
 
+  google.protobuf.BoolValue validateGateway = 4;
```
Comment (hzxuzhonghu) [inline operator/pkg/apis/istio/v1alpha1/values_types.proto:1133]: need a comment here

**Review by hzxuzhonghu — APPROVED** [review]

**Review by costinm — APPROVED** [review]

**Pull request merged**